scripts/gb_model/process_cop_profiles.py

```diff
@@ -0,0 +1,127 @@
+# SPDX-FileCopyrightText: gb-open-market-model contributors
+#
+# SPDX-License-Identifier: MIT
+
+"""
+COP profiles processing for heat pumps.
+
+This script computes the ASHP and GSHP COP profiles weighted by population densities at urban and rural areas and saves it in CSV file format.
+"""
+
+import logging
+from itertools import chain
+from pathlib import Path
+
+import pandas as pd
+import xarray as xr
+
+from scripts._helpers import configure_logging, set_scenario_config
+
+logger = logging.getLogger(__name__)
+
+
+def process_cop_profiles(
+    cop_path: str,
+    clustered_population_layout_path: str,
+    district_heat_share_path: str,
+    year: int,
+    heat_pump_sources: list[str],
+) -> pd.DataFrame:
+    """
+    Process PyPSA-Eur COP profiles across urban and rural areas to obtain ASHP and GSHP profiles weighted by population
+
+    Args:
+        cop_path(str): Filepath to the COP profile generated by existing PyPSA-Eur rules
+        clustered_population_layout_path(str): Filepath to the population density at each node generated by existing PyPSA-Eur rules
+        district_heat_share_path(str): Filepath to district heat share at each country
+        year(int): Modelling year
+        heat_pump_sources (list[str]): List of sources for the heat pumps
+
+    Returns:
+        pd.DataFrame : pandas dataframe containing the weighted average COP profiles
+    """
+    clustered_population_layout = pd.read_csv(
+        clustered_population_layout_path, index_col="name"
+    )
+
+    district_heat_share_data = pd.read_csv(
+        district_heat_share_path, index_col="country"
+    )
+    # Current data exists from 1990-2021, Choosing 2021 to project data for all future years
+    if int(year) >= 1990 and int(year) <= 2021:
+        district_heat_share = district_heat_share_data.loc[:, year]
```

Comment:
The FES has future looking district heat share (e.g. CV.18 has `frac_DH` for 2050)